CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+# Redis Cluster v2 Proposal
```

Comment:
Still no conversation about how to keep `Functions` consistent across the cluster. I think that is still an important item to try to solve. It doesn't solve all that difficult to store it in the TDs, but it will require interleaving functions with replication traffic. (Mixing CP and DP functionality, which isn't great)

Reply:
This is a difficult one.
If we perceive functions (and lua as well prob.) as configuration then it makes sense to use the TD/FC layers to distribute them.

It might be worth getting some acutal usage stats on Lua (doubt there's enough on  Functions atm) - would current usage easily lend itself to a 'configuration' pattern of infrequent changes.

Reply:
Unlike EVAL (which is the user's app responsibility), we should think of Functions as a light weight Modules.
They're an extending the API of the database and should be propagated to all nodes at function registration time and later when new nodes join.
Unlike EVAL, the user's app shouldn't be the one that loads them, but rather the admin, so either the admin tools or the cluster itself, must be the ones responsible to make sure they exist on all nodes.
I suppose that the difference between functions and modules in that respect is that for modules we don't want to offer any code to distribute the binaries to different hosts and load them (we should leave that to some orchestration system that deploys the software and creates the config files), but for functions we want a lower entry barrier.
We need to think about this more, and when we do, let's consider getting ACL into that mix too.

Reply:
If like @oranagra says the envisioned usage is 'configuration/lightweight module' then I don't see an issue with propagating functions via the control path.
The data can be added as a section to the topology sent by the TD with it's own epoch to allow for caching.

Otherwise, we will need to create a different mechanism (perhaps using the same logic as my pubsub handling proposal -- with modifications to guarantee at least once delivery and to handle new nodes).

Reply:
I think propagating by the control path, in this case, is probably what we want. In a lot of ways it is like an external operator going in and calling a command on every node in the cluster and making sure it applied correctly. The expectation should be that functions are low throughput. I would even posit we could be aggressive and say you can only propagate one function at a time, just too artificially slow down the rate of propagation.

Reply:
Treating them as configuration makes a lot of sense to me, but since we don't have sufficient knowledge on how functions will be used in the wild - maybe defer the decision? 
In the next couple of months we may see a more 'lua-ish' usage which would then completely break once we GA Flotilla.

Reply:
maybe we could defer that decision, but i'd like to point out two things:

first, we should be the ones to decide how functions are used in the wild, and adding mechanisms around their propagation could assist. standing aside to see what users would do, is likely to promote a misuse.

of course in the context of Flotilla, the above statement is not relevant since it's not yet available to users, so indeed we can take the time. (i just like to avoid taking a similar approach elsewhere).

secondly, arguably sooner or later, we'll need to propagate some (non-cluster related) configuration (be it ACL, Functions or something else) through the cluster, so i think it may be better to include this topic in the design in an early stage.

Reply:
Agree wrt ACLs and other configs - will add to the design.

Reply:
+1 on adding a section for propagating configuration in general.